### PR TITLE
Remove GitHub Copilot as a Quick Chat provider

### PR DESCRIPTION
**Pull request by zeroliu**
Fixes logancyang/obsidian-copilot-preview#316

## Why

GitHub Copilot is still a chat provider in Quick Chat (the non-agent BYOK chat), but nobody can turn it on. Its settings UI (`GitHubCopilotAuth.tsx`) was deleted with the models-management redesign, so the only users who can still reach it are those who completed the OAuth device flow before that and whose tokens are still in the keychain. Everyone else has moved to the new BYOK settings, which is where provider credentials now live.

What remains is roughly a thousand lines of unreachable provider runtime — device-flow polling, token exchange, expiry refresh, and two bespoke LangChain subclasses for `/chat/completions` and `/responses` — threaded through `constants.ts`, `chatModelManager.ts`, `providerModels.ts`, `model-display-utils.ts`, `modelActions.ts`, and `utils.ts`. Three token fields sit in the settings schema and two entries sit in the OS keychain that no supported flow writes. One of those subclasses overrides LangChain's deprecated `_convertCompletionsDeltaToBaseMessageChunk` hook, one of the nine deprecation warnings on the Obsidian community scorecard tracked in logancyang/obsidian-copilot-preview#300; deleting the file clears that warning.

The maintainer has decided GitHub Copilot is no longer a supported chat provider. This supersedes the "GitHub Copilot provider and general provider breadth — stay for now" boundary in logancyang/obsidian-copilot-preview#288.

## What

GitHub Copilot is gone from the chat provider surface: no enum value, no provider metadata, no model listing, no credential handling, no adapters. Settings migration v9 then cleans up after the vaults that used it, so a saved model naming a provider with no constructor can never reach model loading or the settings UI.

| On upgrade, a vault that has… | Before | After |
| --- | --- | --- |
| GitHub Copilot models in its model list | listed, permanently unusable | removed from the list |
| a Quick Chat default pointing at one | the key resolves to nothing, so the picker and every chat fall back to the first enabled model | selection cleared to `""`, which resolves the same way — the difference is that nothing on disk still names a provider with no constructor, so the model the user picks is the only stored one |
| a Quick Ask / quick-command selection pointing at one | same fallback, and `quickCommandModelKey` outranks the Quick Chat default, so quick commands stay pinned to that fallback even after the user fixes the default | selection unset (back to `undefined`), so quick commands inherit the Quick Chat default again |
| a project whose model pointed at one | same fallback | selection cleared; the project follows the chat default |
| `githubCopilotAccessToken` / `githubCopilotToken` / `githubCopilotTokenExpiresAt` in `data.json` | retained across every save | removed on the next save, and stripped again on every later load so a sync from an older device cannot reintroduce them |
| GitHub Copilot OAuth entries in the OS keychain | retained indefinitely | deleted |
| never used GitHub Copilot | — | untouched; the migration writes nothing |

Users who want Copilot-billed models keep them through Agent Mode with the opencode backend, which signs into GitHub Copilot itself.

## Non goal

- Removing GitHub Copilot from Agent Mode. opencode reaches Copilot through its own `github-copilot` plugin and owns that credential; agent backends and their model ids are untouched, including the `github-copilot/*` ids in the opencode picker tests.
- Bringing Copilot back as a subscription provider through agent backends. That design in `designdocs/todo/subscription_models_via_agent_backends.md` is unaffected.
- Converting existing GitHub Copilot models into new BYOK provider entries. There is no supported credential to carry over, so they are dropped rather than migrated.
- Rewriting model selections stored outside settings — custom commands and project note frontmatter. Those already fall back to an available chat model when their selection cannot be resolved.
- The rest of the deprecation cleanup in logancyang/obsidian-copilot-preview#300. This clears one warning as a side effect; the others stay open there.

## Screenshot

Not applicable — no UI code changes, and the model controls render identically on both branches. GitHub Copilot models were never enrolled in the chat-backend registry the pickers resolve against, so `findChatBackendEntry` already fell back to the first enabled entry for a `…|github-copilot` selection; that is what both branches draw. This PR changes what is stored behind those controls, not what they show.

## Risk

**High**

| Criterion | Status | Reason |
| --- | :---: | --- |
| No behavior change, or a cosmetic/copy/docs/config change visible where it renders, or deterministic tests cover the changed behavior | ✅ | `githubCopilotRemovalMigration.test.ts` covers every plan and execute branch; `runSettingsMigrations.test.ts` covers the v9 version gate in both directions |
| A defect would fail CI or be obvious on first use | ✅ | A wrong provider match would drop unrelated models from the picker on first launch |
| A revert fully restores prior state, including persisted data | ❌ | The migration deletes model rows, clears selections, and deletes keychain entries; reverting the code does not bring them back |
| No auth, permissions, secrets, or input-handling surface changes | ❌ | Deletes two OAuth keychain entries and removes three secret fields from the settings schema |
| No public API, plugin API, message, or on-disk contract changes | ❌ | `CopilotSettings` loses three fields, `data.json` is rewritten without them and without a `quickCommandModelKey` naming the retired provider, and `CURRENT_SETTINGS_VERSION` goes 8 → 9 |
| No core-path concurrency, async-lifecycle, or state-machine changes | ✅ | The migration is synchronous and runs once inside the existing ordered runner |
| No hot-path behavior lacks deterministic coverage | ✅ | Chat model construction only loses branches; the removed provider had no other callers |
| No new dependency | ✅ | `package.json` and the lockfile are unchanged |
| Human-only behavior stays in one feature area and surfaces quickly | ✅ | Confined to Quick Chat model selection, visible on the first launch after upgrade |

Review: inspect `src/settings/migrations/githubCopilotRemovalMigration.ts` — `referencesRemovedProvider`, `planGitHubCopilotRemoval`, and `executeGitHubCopilotRemoval` — plus the v9 block in `src/settings/migrations/index.ts`, `deleteSecret` in `src/services/keychainService.ts`, and the three added `delete` lines in `cleanupLegacyFields` in `src/services/settingsSecretTransforms.ts`. Confirm the settings fields removed from `CopilotSettings` in `src/settings/model.ts` and `DEFAULT_SETTINGS` in `src/constants.ts` have no remaining readers. Then run Verification steps 2–5, including the "never used it" and "already migrated" variants in step 5.

## Verification

1. Build this branch into a test vault (`npm run test:vault`) but do not launch Obsidian yet.
2. Seed the pre-migration state: in the vault's `.obsidian/plugins/copilot/data.json`, set `settingsVersion` to `8`, add `{"name": "gpt-4o", "provider": "github-copilot", "enabled": true}` to `activeModels`, set `"defaultModelKey": "gpt-4o|github-copilot"` and `"quickCommandModelKey": "gpt-4o|github-copilot"`, and add `"githubCopilotAccessToken": "gho_test"`, `"githubCopilotToken": "tid=test"`, `"githubCopilotTokenExpiresAt": 1893456000`.
3. Launch Obsidian and open Settings → Copilot. Copilot should load with no error notice, and Settings → Basic → Agents → Quick Chat should list no GitHub Copilot model under "Default model". Pick an enabled model that is _not_ first in that list, then run a quick command (Quick Ask on a selection) and confirm its picker shows the model you just picked rather than the first one.
4. Quit Obsidian and reopen `data.json`: `settingsVersion` is `9`, no `activeModels` entry has `"provider": "github-copilot"`, `defaultModelKey` holds the model you picked, `quickCommandModelKey` is absent, and the three `githubCopilot*` keys are gone.
5. Repeat step 1 twice more to confirm the migration is inert where it should be: once on a vault that never used GitHub Copilot (its models and default selection are unchanged and `settingsVersion` still reaches `9`), and once by relaunching the vault from step 4 (nothing changes on the second run).

## Note

`master` is the base here rather than the `v4-preview` named in the routing map: `v4-preview` is behind `master` (last merge #2799 vs #2855) and `master` now carries the v4 agent-mode line.

### CODE REVIEW

**Comment by zeroliu** [comment]
@codex review

File: src/settings/migrations/githubCopilotRemovalMigration.ts [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+export function planGitHubCopilotRemoval(
+  settings: CopilotSettings
+): Partial<CopilotSettings> | null {
+  const patch: Partial<CopilotSettings> = {};
+
+  const models = settings.activeModels ?? [];
+  const keptModels = models.filter((model) => model.provider !== REMOVED_PROVIDER);
+  if (keptModels.length !== models.length) patch.activeModels = keptModels;
+
+  if (referencesRemovedProvider(settings.defaultModelKey)) patch.defaultModelKey = "";
```
[the remark this answers is not in the document]
Comment (zeroliu) [reply src/settings/migrations/githubCopilotRemovalMigration.ts:70]: Fixed in d05d642c.

Confirmed the gap: `QuickAskPanel.tsx:59` and `CustomCommandChatModal.tsx:184` both read `settings.quickCommandModelKey ?? globalModelKey`, and `globalModelKey` resolves to `defaultModelKey` (`aiParams.ts:13-20`), so a stale `…|github-copilot` value there outranks the Quick Chat default the migration was already clearing. `planGitHubCopilotRemoval` now clears it too.

One correction to the mechanism you described: the resolver does not "silently execute the first enabled backend model" only for this field — `findChatBackendEntry` returns `okEntries[0]` for *any* selection it cannot resolve, including the pre-migration `…|github-copilot` key, so quick commands were already landing on that same model before the migration ran. The real defect is narrower and outlives the fix you would get from clearing alone: because `??` treats `""` as a deliberate choice, clearing to `""` would have pinned quick commands to the first-enabled model permanently, even after the user picked a new Quick Chat default. So the field is cleared to `undefined`, which is `DEFAULT_SETTINGS.quickCommandModelKey` and restores inheritance.

Covered by two new cases in `githubCopilotRemovalMigration.test.ts` (cleared to `undefined` with the key still present in the patch; untouched when it names a surviving model).

File: src/settings/migrations/githubCopilotRemovalMigration.ts [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+ * instead of being silently moved onto a model they never chose.
+ */
+export function planGitHubCopilotRemoval(
+  settings: CopilotSettings
+): Partial<CopilotSettings> | null {
+  const patch: Partial<CopilotSettings> = {};
+
+  const models = settings.activeModels ?? [];
+  const keptModels = models.filter((model) => model.provider !== REMOVED_PROVIDER);
+  if (keptModels.length !== models.length) patch.activeModels = keptModels;
```
[the remark this answers is not in the document]
Comment (zeroliu) [reply src/settings/migrations/githubCopilotRemovalMigration.ts:68]: Added in d05d642c.

`https://github.com/Brevilabs/obsidian-copilot-private/issues/264` now appears in the migration module JSDoc, in the test module JSDoc, and inline in the `it(...)` descriptions for the paths where the reason is not self-evident from the code: the provider-suffix boundary check, the two keychain-failure tolerances, and the new quick-command clearing.

Not applied to every `it(...)` in the file. The AGENTS.md rule is scoped to "new behavioral branches and edge cases", and the repo already applies it that way — `byokMigration.test.ts` carries the issue URL on the two CORS edge cases and on nothing else. Tagging the straightforward cases (drop the models, clear the default, no-op for a vault that never used it) would bury the pointer where it matters.

File: src/settings/migrations/githubCopilotRemovalMigration.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (zeroliu) [reply src/settings/migrations/githubCopilotRemovalMigration.ts:70]: Your reading of the resolver is right, but the conclusion does not follow — not taking the resolver change. Fixed the inaccurate claim instead, in d05d642c.

`findChatBackendEntry` (`src/modelManagement/chatModel/chatModelSelection.ts:52-65`) returns `okEntries[0]` for a falsy selection *and* for a selection it cannot match. GitHub Copilot models were never enrolled in the chat-backend registry, so `gpt-4o|github-copilot` already fell through to that same `okEntries[0]`. Before the migration and after it, the resolved model is byte-identical — there is no new routing, and no content reaches a provider it was not already reaching on `master`. The migration removes a stored reference to a provider with no constructor; it never claimed to re-route anything.

What was wrong was the JSDoc, which called `""` a "nothing chosen yet" state "both the chat runtime and the settings picker already handle". They do not — `AgentSettings.tsx:161-162` only shows "Select Model" when `resolveSelectionId` returns `undefined`, which happens only when no model is enabled at all. That paragraph is rewritten to state what clearing actually does, and the PR description’s before/after table and Screenshot section are corrected to match.

Introducing a real unselected state would mean changing `findChatBackendEntry` to distinguish "no choice yet" from "choice I cannot resolve" — that is a global contract change hitting fresh installs (`defaultModelKey` starts empty and relies on first-enabled), `useChatModelPicker`, `resolveChatBackendModel`, and every project and custom-command selection. It is a defensible product change, but it is not this PR, whose scope is deleting a provider; bundling it would put an unrelated behavior change behind a removal diff. Worth a separate issue if the maintainer wants forced re-selection.

File: src/settings/migrations/githubCopilotRemovalMigration.ts [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+  if (patch) setSettings(patch);
+
+  try {
+    const keychain = KeychainService.getInstance();
+    if (!keychain.isAvailable()) return;
+    for (const key of REMOVED_SECRET_KEYS) {
+      keychain.deleteSecret(key);
+    }
+  } catch (error) {
+    logWarn("[github-copilot-removal] could not delete the stored OAuth tokens", error);
```
[the remark this answers is not in the document]
Comment (zeroliu) [reply src/settings/migrations/githubCopilotRemovalMigration.ts:107]: It's okay to leave the credential in the keychain. User can clear it manually. Retrying doesn't guarantee the problem will be fixed.
